cores/esp32/IPAddress.h

```diff
@@ -26,55 +26,60 @@
 
 // A class to make it easier to handle and pass around IP addresses
 
+#define IPADDRESS_V4_BYTES_INDEX 12
+#define IPADDRESS_V4_DWORD_INDEX 3
+
+enum IPType
+{
+    IPv4,
+    IPv6
+};
+
 class IPAddress: public Printable
 {
 private:
     union {
-        uint8_t bytes[4];  // IPv4 address
-        uint32_t dword;
+        uint8_t bytes[16];
+        uint32_t dword[4];
     } _address;
+    IPType _type;
```

Comment:
Would it be possible to use `ip_addr_t` instead of this local structure. This is what is done in ESP8266 and makes it easier to call esp-idf functions, saving conversions from `IPAddress` and `ip_addr_t`

Reply:
Please also initialize `_type` as it is not always initialized. (e.g. when using `fromString4` or `fromString6`)
So maybe the type could be extended to have a 3rd type: `NotSet` or something similar.
But then the checks for the `_type` should include both checks, for IPv4 and IPv6 as now it is more like a boolean check. It is either IPv4 or IPv6.

Reply:
Esp-idf has a 'not set' IP although I'm not a big fan because it creates confusion to users and needs additional tests. 

I'm more in favor of having IP ADDRESS always initialized to ANY to the widest supported length, I.e. v6 of `LWIP_IPV6` is defined

Reply:
> `_type` as it is not always initialized

fromString is an instance method, not static/factory. By default _type is initialized to IPv4, to be backwards compatible with existing code. Using xxx(IPv6) initialises to IN6ADDR_ANY.

(Which I noticed setting the constant was missing / not copied from ArduinoCore-API).

If you then call fromString() and it successfully parses as either 4 or 6 then it will be set accordingly.

If it fails to parse, then it will remain as is (usually IPv4), but the byte values will be meaningless (partially parsed).

Maybe if parse fails, then it should set to all zeros? (which will be IPv4 ANY or IPv6 ANY, depending on the original)

Reply:
@sgryphon I guess you're right.

Missed that one. I haven't run into situations where the _type variable was uninitialized, just looked through the code to see if all were initialized as I've been bitten by uninitialized variables myself quite a lot lately and thus was looking through the code to see if all were initialized.

> Maybe if parse fails, then it should set to all zeros? (which will be IPv4 ANY or IPv6 ANY, depending on the original)

I guess it should be set to some default constructed state, where it would be obvious to take the default IPv4 constructor for the `fromString4` and IPv6 for the `fromString6`.

Then calls like `isSet` (or whatever the call is for `IPAddress`) will return false.

Reply:
Ah yes, so the type itself should have an any type as well. (or maybe a bitmask, for IPv4 any and IPv6 any?, but this will break LWIP consistency, so maybe not...)
So I guess it should just not change the address in any way if the `fromStringX` call was not given a properly formatted string?

Reply:
I added a suggestion for this in the comments around the ANY constants.